---
title: "E03 Statistics for popularion genetics 2"
author: "Urška Gerič"
date: "2024-11-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r message=FALSE, results='hide', warning=FALSE}
library(dartR)
library(adegenet)
library(SNPRelate)
library(HardyWeinberg)
library(ggtern)
library(dartRverse)
library(reshape2)
library(seqinr)
library(MultiPhen)
library(tidyverse)
library(pegas)
library(hierfstat)
```

If you don't have the packages installed, you can install them by running the following code:

```{r}
#install.packages("adegenet")

#if (!require("BiocManager", quietly = TRUE))
  #install.packages("BiocManager")

#BiocManager::install("dartR")
#BiocManager::install("SNPRelate")
#install.packages("HardyWeinberg")
#install.packages("ggtern")
#install.packages("dartRverse")
```

# Population genetics

In population genetics we can use many different parameters and statistics to describe our data. We will use some of them.


# Adegenet package

## Dataset nancycats

```{r}
data(nancycats)

#locus names
locNames(nancycats)
```


```{r}
#populations 
pop(nancycats)
```

Number of individuals in each population

```{r}
table(pop(nancycats))
```

Some basic statistics

```{r}
stats <- summary(nancycats)
stats
```

### Plots

```{r}
par(mfrow=c(2,2))

plot(stats$n.by.pop, 
     stats$pop.n.all, 
     xlab="Colonies sample size",
     ylab="Number of alleles",
     main="Alleles numbers and sample sizes",
     type="n")
text(stats$n.by.pop,stats$pop.n.all,lab=names(stats$n.by.pop))

barplot(stats$loc.n.all, 
        ylab="Number of alleles",
        main="Number of alleles per locus")

barplot(stats$Hexp-stats$Hobs, 
        main="Heterozygosity: expected-observed",
        ylab="Hexp - Hobs")

barplot(stats$n.by.pop, 
        main="Sample sizes per population",
        ylab="Number of genotypes",las=3)
```


```{r}
par(mfrow=c(1,1))
```

### Hardy-weinberg equilibrium

```{r}
cats.hwt <- hw.test(nancycats, B=0)
cats.hwt
```

### F statistics

```{r}
#global FST and FIS
wc(nancycats)
```

F statistics per locus

```{r}
fst_cats <- Fst(as.loci(nancycats))
fst_cats
```

Mean F statistics

```{r}
colMeans(fst_cats)
```

F statistics with package `hierfstat`

```{r}
hier_cats <- genind2hierfstat(nancycats)
boot.vc(hier_cats[1], hier_cats[-1])$ci
```

### Genetic distances

We can calculate genetic distances between populations with Nei distance.

```{r}
fst_mat <- genet.dist(nancycats[1:50, ], method="Nei87")
fst_mat
```


```{r}
is.euclid(fst_mat)
```

### Inbreeding 

```{r}
inb_cats <- inbreeding(nancycats)

fbar <- sapply(inb_cats, function(x) mean(x, na.rm=T))

hist(fbar, 
     col="lightblue", 
     main="Distribution of inbreeding coefficients", 
     xlab="Inbreeding coefficient")

```

### glplot

For genlight objects 

```{r message=FALSE, results='hide', warning=FALSE}
df_hapmap <- read.PLINK("hapmap_A_raw.raw", 
                            map.file = "hapmap_A.map", 
                            ped.file  = "hapmap_A.ped",
                            type = "genind")

```


```{r}
glPlot(df_hapmap, posi = "topleft")
```

## Missing values

```{r}
missval <- glNA(df_hapmap)

hist(missval, proba=T, col = "lightblue", main = "Distribution of missing data per locus", xlab = "Frequency of missing data per locus")
```

We can see that we can do a lot of things with adegenet, but the problem is convert raw plink data to genind object. We can convert it to genlight object which has fewer functions in adegenet package. Alternative to adegenet is dartR package.

# DartR package

Load the data 

```{r message=FALSE, warning=FALSE, results='hide'}
dogs <- read.PLINK(file="NordicDogs.raw", map.file = "NordicDogs.map.map")
```

This command transform plink files into a genlight object. Genlight object have data about SNPs.

Here we can see number of SNPs, individuals, genotypes and so on:

```{r}
dogs
```

We can get information about populations:

```{r}
dogs@pop
```

Data about individuals(names):

```{r}
dogs$ind.names
```


We can see that we have really big dataset, so we will subset it to make it easier to work with. We will keep only 9000 SNPs.

```{r}
dogs_subset <- gl.keep.loc(dogs, first=1001, last=2000)
dogs_subset
```

We now need to check if our dataset is ok for using dartR:

```{r}
dogs_subset <- gl.compliance.check(dogs_subset)
```

## Allele frequencies

```{r}
allele_freq <- gl.alf(dogs_subset[, 1:10])
```

We can now plot allele frequencies. We need to transform our dataframe and make it as matrix.

```{r}
barplot(t(as.matrix(allele_freq)), 
        xlab="Alleles", 
        ylab="Frequency", 
        main="Allele frequencies")
```

## Heterozygosity

```{r}
gl.report.heterozygosity(dogs_subset)
```


The output displays heterozygosity by populations (Ho - observed, He - expected) and FIS (Inbreeding Coefficient). Ho and He values range from 0 to 1, where 1 represents 100% heterozygosity. FIS values range from -1 to 1, where positive FIS values indicate that individuals in the population are more related than expected under a random mating model, and negative FIS values indicate that individuals in the population are less related than expected under a random mating model.

## Hardy-Weinberg equilibrium

Ternary plots illustrating the patterns of Hardy-Weinberg (HW) proportions. Each vertex on the plot represents a different genotype: homozygous for the reference allele (AA), heterozygous (AB), and homozygous for the alternative allele (BB). The plots highlight loci that deviate significantly from Hardy-Weinberg equilibrium, and these loci are indicated in pink. The blue parabola on each plot represents Hardy-Weinberg equilibrium, while the area between the green lines represents the acceptance zone. The plots provide a visual representation of the distribution of the SNPs in relation to the Hardy-Weinberg equilibrium and allow for the identification of loci that may be under selection or experiencing other evolutionary forces.

Make ternary plots for each population:

```{r}
#subset
gl.report.hwe(dogs_subset, subset = "each")

#whole dataset
gl.report.hwe(dogs, subset = "each")
```

Make ternary plot for whole dataset:

```{r}
gl.report.hwe(dogs_subset, subset = "all")
```

Make heatmap of Hardy-Weinberg equilibrium p-values:

```{r}
#subset
gl.hwe.pop(dogs_subset)

#whole dataset
gl.hwe.pop(dogs)
```


## FST

With FST value we can measure genetic distance between populations. FST indicates the likelihood that two randomly selected genes from two populations share a common origin.
If we use the argument "nboots," the program performs 100 bootstrap repetitions to calculate p-values.

```{r results='hide', message=FALSE, warning=FALSE}
fst <- gl.fst.pop(dogs_subset, nboots=100)
fst$Fsts
```

Three matrices are displayed:

- The first matrix provides FST values between populations, where higher values indicate a higher level of genetic differentiation between populations.
- The second matrix provides p-values.

If the p-value is below 0.05, the difference is statistically significant.

## Heatmap of FST values:

First we need to transform our data to dataframe:

```{r}
fst_df <- melt(as.matrix(fst$Fsts))
colnames(fst_df) <- c("Population1", "Population2", "FST")
head(fst_df)
```

Plot heatmap:

```{r}
# Plot heatmap using ggplot2
ggplot(fst_df, aes(x = Population2, y = Population1, fill = FST)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red", name = "FST") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Pairwise FST Heatmap", x = "Population 1", y = "Population 2")
```


## Effective population size

The effective population size can be estimated using the gl.LDNe() function. This function estimates Ne based on linkage disequilibrium (LD).

We say that loci are in LD when the frequency of association between their alleles is higher or lower than expected if the loci were independent and associated randomly.
For example, if two loci are located close to each other, they are usually inherited together.
The function also requires the Ne2-1 (L, M, .exe) file, depending on your operating system.

```{r}
ne_sub <- gl.LDNe(dogs_subset)
ne_all <- gl.LDNe(dogs)
ne_all
```


# Export results in csv files

```{r results='hide', message=FALSE, warning=FALSE}
write.csv(allele_freq, "allele_freq.csv")
write.csv(gl.report.heterozygosity(dogs_subset), "heterozygosity.csv")
write.csv(fst$Fsts, "fst.csv")
write.csv(fst_df, "fst_df.csv")
write.csv(fst$Bootstraps, "fst_boot.csv")
```